node_heap.go

```diff
@@ -53,9 +53,14 @@ func (nh *NodeHeap) TopN(n int) []*Node {
 	m := make([]*Node, 0, n)
 	nh.lk.RLock()
 	defer nh.lk.RUnlock()
-	for i := 0; i < n && i < len(nh.Nodes); i++ {
-		node := nh.Nodes[i]
-		m = append(m, node)
+	for i := 0; i < n; i++ {
```

Comment:
@willscott noticed a slight issue with the use of heaps here. 

The heap will keep itself sorted upon insertion and removal of new nodes. However, the existing nodes in the heap have dynamically changing values and the heap does not continuously sort itself every time one of those values changes. 

The strategy I've implemented here is to select the top N nodes from the heap, to use `.Pop()` n times then to retrieve the top nodes, then add them back using `.Push()`. This is a similar flow to the `.Fix()` method that exists on the heap which is meant to reposition an item in the heap if it changes its value. Not the most efficient, but we are dealing with fairly small heap sizes here. 

Also wondering if we still think that a priority queue is the best data structure to store the `AllNodes` tier? My understanding of the code / implementation is that the `ActiveNodes` is always a subset of `AllNodes` and they are not mutually exclusive sets like the previous "Main" and "Uknown" tier design.

Reply:
@willscott One issue I noticed with my changes is that since the nodes are temporarily removed from the heap, if there are any stats being written to them concurrently, that will fail.

Reply:
what do you mean by 'fail'? a temporary dropped metric isn't the end of the world, potentially.

If not a priority queue for all nodes, how would you imagine keeping track of the whole set?

Reply:
one note on golang semantics here: you have the `nh.lk.RLock()` above - that indicates you're taking a "read only lock" over the data structure - but the pop and push you're proposing in your change will do a read-write operation.

Reply:
Thanks for the tip here Will, always good to learn :)